File: projects/Vibe_XASR/windows_build/src/VibeXASR.Windows/Lexicon/HotwordsStore.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace VibeXASR.Windows.Lexicon;

/// <summary>
/// Persists the user's hotword list to a plain-text file sherpa-onnx loads (one phrase per line).
/// Faithful port of the macOS HotwordsStore: normalize (trim, drop blanks + '#' comments),
/// space-separate CJK chars (this model has no bare-char tokens — each CJK char maps to its own
/// ▁-piece via the augmented bpe.vocab), per-word boost (CJK uses the full score; pure-English is
/// capped ≤2.5 and ALSO emitted capitalized-first since the model only emits capitalized pieces),
/// atomic write. Writing an empty list removes the file so the engine stays on greedy_search
/// (zero behaviour change).
/// </summary>
public static class HotwordsStore
{
    private static readonly Regex ScoreSuffix = new(@"\s:\d+(\.\d+)?$", RegexOptions.Compiled);

    public static List<string> Normalize(string? text) =>
        (text ?? "").Split('\n', '\r')
            .Select(l => l.Trim())
            .Where(l => l.Length > 0 && !l.StartsWith("#"))
            .ToList();

    public static int Count(string? text) => Normalize(text).Count;

    private static bool IsCjk(char c) =>
        (c >= 0x3400 && c <= 0x4DBF) || (c >= 0x4E00 && c <= 0x9FFF) || (c >= 0xF900 && c <= 0xFAFF);

    /// <summary>Space-separate every CJK char; English words intact; trailing " :score" preserved.
    /// e.g. "我用OpenAI" → "我 用 OpenAI", "李沐 :2.5" → "李 沐 :2.5".</summary>
    public static string SpaceCjk(string s)
    {
        var sb = new StringBuilder(s.Length * 2);
        foreach (var ch in s)
        {
            if (IsCjk(ch)) sb.Append(' ').Append(ch).Append(' ');
            else sb.Append(ch);
        }
        return string.Join(" ", sb.ToString().Split(' ', StringSplitOptions.RemoveEmptyEntries));
    }

    private static string FmtScore(double s) =>
        s == Math.Round(s) ? ((int)s).ToString(CultureInfo.InvariantCulture)
                           : s.ToString("0.###", CultureInfo.InvariantCulture);

    /// <summary>Append a per-word boost: CJK terms use the full score; pure-English is capped ≤2.5
    /// (over-boosting English distorts words). An explicit trailing " :N" is respected as-is.</summary>
    public static string WithBoost(string line, double score)
    {
        var m = ScoreSuffix.Match(line);
        if (m.Success)
        {
            var word = line.Substring(0, m.Index);
            var suffix = line.Substring(m.Index).Trim();
            return SpaceCjk(word) + " " + suffix;
        }
        double s = line.Any(IsCjk) ? score : Math.Min(score, 2.5);
        return SpaceCjk(line) + " :" + FmtScore(s);
    }

    public static string CapitalizedFirst(string s) =>
        s.Length == 0 ? s : char.ToUpperInvariant(s[0]) + s.Substring(1);

    /// <summary>Expand a raw line into prepared line(s). A pure-English word ALSO emits a
    /// capitalized-first variant (the model emits "▁Py", never "▁py").</summary>
    public static List<string> Expand(string line, double score)
    {
        if (ScoreSuffix.IsMatch(line) || line.Any(IsCjk)) return new() { WithBoost(line, score) };
        var cap = CapitalizedFirst(line);
        var variants = cap == line ? new List<string> { line } : new List<string> { line, cap };
        return variants.Select(v => WithBoost(v, score)).ToList();
    }

    /// <summary>Write the prepared list to <paramref name="path"/>; remove it when empty. Returns
    /// true if a non-empty file was written.</summary>
    public static bool WriteFile(string? text, double score, string path)
    {
        var lines = Normalize(text).SelectMany(l => Expand(l, score)).ToList();
        try
        {
            if (lines.Count == 0) { if (File.Exists(path)) File.Delete(path); return false; }
            Directory.CreateDirectory(Path.GetDirectoryName(path)!);
            File.WriteAllText(path, string.Join("\n", lines) + "\n", new UTF8Encoding(false));
            return true;
        }
        catch (Exception ex) { Diag.Log("hotwords write failed: " + ex.Message); return false; }
    }

    public static bool IsNonEmpty(string path)
    {
        try { return File.Exists(path) && new FileInfo(path).Length > 0; }
        catch { return false; }
    }
}

File: projects/Vibe_XASR/macos_build/native/app/Sources/VibeIME/HotwordsStore.swift
import Foundation

/// Persists the user's hotword list to a plain-text file the sherpa-onnx engine
/// loads (one phrase per line). Normalizes on write: trims each line, drops blank
/// lines and `#` comments. Writing an empty list removes the file so the engine
/// falls back to greedy_search with no biasing (zero behaviour change).
///
/// Format notes (sherpa-onnx hotwords-file):
///   * one phrase per line, UTF-8, Chinese / English / mixed written verbatim;
///   * a trailing ` :2.5` on a line overrides the per-word boost.
enum HotwordsStore {

    /// Split the raw editor text into normalized hotword lines.
    static func normalize(_ text: String) -> [String] {
        text.split(whereSeparator: \.isNewline)
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty && !$0.hasPrefix("#") }
    }

    /// Number of effective hotwords in the raw text (for the UI counter).
    static func count(_ text: String) -> Int { normalize(text).count }

    private static func isCJK(_ c: Character) -> Bool {
        guard c.unicodeScalars.count == 1, let v = c.unicodeScalars.first?.value else { return false }
        return (0x3400...0x4DBF).contains(v) || (0x4E00...0x9FFF).contains(v) || (0xF900...0xFAFF).contains(v)
    }

    /// Space-separate every CJK character so sherpa's BPE tokenizer maps each to
    /// its own ▁-prefixed piece (this model has no bare-char tokens). English
    /// words are left intact; a trailing " :score" boost suffix is preserved.
    /// e.g. "我用OpenAI" → "我 用 OpenAI", "李沐 :2.5" → "李 沐 :2.5".
    static func spaceCJK(_ s: String) -> String {
        var out = ""
        for ch in s {
            if isCJK(ch) { out += " "; out.append(ch); out += " " } else { out.append(ch) }
        }
        return out.split(separator: " ").joined(separator: " ")
    }

    /// Format a score compactly (5.0 → "5", 2.5 → "2.5").
    private static func fmtScore(_ s: Double) -> String {
        s == s.rounded() ? String(Int(s)) : String(format: "%g", s)
    }

    /// Per-word boost so ONE UI strength works for both scripts: CJK terms are
    /// char-level and need a stronger push (`score`); pure-English terms are
    /// capped lower (≤2.5) because over-boosting English distorts words. An
    /// explicit trailing " :N" the user typed is respected as-is.
    static func withBoost(_ line: String, score: Double) -> String {
        if let r = line.range(of: #"\s:\d+(\.\d+)?$"#, options: .regularExpression) {
            let word = String(line[..<r.lowerBound])
            let suffix = String(line[r.lowerBound...]).trimmingCharacters(in: .whitespaces)
            return spaceCJK(word) + " " + suffix
        }
        let s = line.contains(where: isCJK) ? score : min(score, 2.5)
        return spaceCJK(line) + " :" + fmtScore(s)
    }

    /// Capitalize just the first character ("pytorch" → "Pytorch").
    static func capitalizedFirst(_ s: String) -> String {
        guard let f = s.first else { return s }
        return f.uppercased() + String(s.dropFirst())
    }

    /// Expand a raw hotword line into the prepared line(s) written to disk. For a
    /// pure-English word we ALSO emit a capitalized-first variant, because the
    /// model only emits capitalized proper-noun pieces (e.g. "▁Py" in "PyTorch")
    /// and never lowercase "▁py", so a lowercase entry like "pytorch" would match
    /// nothing. CJK / explicit-boost lines pass through unchanged.
    static func expand(_ line: String, score: Double) -> [String] {
        let hasExplicit = line.range(of: #"\s:\d+(\.\d+)?$"#, options: .regularExpression) != nil
        if hasExplicit || line.contains(where: isCJK) {
            return [withBoost(line, score: score)]
        }
        let cap = capitalizedFirst(line)
        let variants = (cap == line) ? [line] : [line, cap]
        return variants.map { withBoost($0, score: score) }
    }

    /// Write the normalized list to `url`, appending a per-word boost derived from
    /// `score` (and English case variants). Removes the file when the list is
    /// empty. Returns true when a non-empty file was written.
    @discardableResult
    static func writeFile(text: String, score: Double, to url: URL) -> Bool {
        let lines = normalize(text).flatMap { expand($0, score: score) }
        let fm = FileManager.default
        guard !lines.isEmpty else {
            try? fm.removeItem(at: url)
            return false
        }
        let body = lines.joined(separator: "\n") + "\n"
        do {
            try body.write(to: url, atomically: true, encoding: .utf8)
            return true
        } catch {
            FileHandle.standardError.write(
                "[VibeIME] hotwords write failed: \(error)\n".data(using: .utf8)!)
            return false
        }
    }

    /// True if the file exists and is non-empty.
    static func isNonEmpty(_ url: URL) -> Bool {
        guard let sz = (try? FileManager.default.attributesOfItem(atPath: url.path))?[.size] as? Int
        else { return false }
        return sz > 0
    }
}
